### ISSUE TITLE
Restructure generate_trajectories: batch N samples per subject into single pass

### ISSUE THREAD

**Issue by mmcdermott** [mmcdermott/MEDS_EIC_AR#89]
## Summary

When generating `N_trajectories_per_task_sample` samples per subject, `MEICAR_generate_trajectories` currently runs a **full pass over the dataloader once per sample** (`src/MEDS_EIC_AR/__main__.py:143-172`). If you're generating 20 samples per subject across 3 splits, that's 60 full predict passes, each of which re-prefills every subject's input from scratch and never shares work between passes.

A much better shape — on any backend, including the current HF one — is to **do one pass, with each subject contributing `N` rows to the batches instead of one**. Same-subject rows end up adjacent in batches, which gives us three distinct wins:

1. **On prefix-caching backends (vLLM / SGLang, see #87):** the KV cache for a subject's input is built once and reused across all `N` samples of that subject in the same batch, instead of being re-prefilled `N` times across `N` separate passes. This is the single biggest generation-time speedup available to us once we have such a backend — potentially `N×` on the decode-heavy path.
2. **On any backend including HF:** when a batch contains many copies of the same subject (or same-length subjects), **padding tightens dramatically**. In the limit where `batch_size == N`, every row in a batch is the same length and there is zero padding waste. Current behavior batches heterogeneous subjects and pads to the max-length in each batch, so short subjects pay a padding overhead that grows with the length-variance of the sample.
3. **One iteration of per-pass setup instead of N.** Lightning `trainer.predict` initialization, dataloader worker spawning, GPU warmup, hash-seed reseeding — all of these happen 60 times today on a 3-split × 20-sample run. Amortized, this isn't enormous, but it's non-zero, and it's entirely avoidable.

This issue is **independent of #87**: it's pure work on the CLI / data-layer and doesn't require any backend changes. It's worth doing on its own for the padding win (2) and the setup-amortization win (3), and it's a **prerequisite multiplier** for the SGLang backend work in #87 because without it most of SGLang's prefix-cache advantage is wasted.

## Current state

From `src/MEDS_EIC_AR/__main__.py:143-172`:

```python
for split in inference.generate_for_splits:
    dataloader = ...  # full dataset for this split

    for sample in range(inference.N_trajectories_per_task_sample):
        out_fp = Path(cfg.output_dir) / split / f"{sample}.parquet"
        ...
        seed = hash_based_seed(cfg.get("seed", None), split, sample)
        seed_everything(seed, workers=True)
        predictions = trainer.predict(model=M, dataloaders=dataloader)    # ← full pass
        predictions_df = format_trajectories(dataloader.dataset, predictions)
        pq.write_table(...)
```

Each `trainer.predict(...)` call iterates the entire dataset. For `N = 20` samples per subject, **every subject's input tensor is materialized and prefilled 20 times** — once per outer-loop iteration — with no sharing between iterations.

### High-level

Replace the nested `for sample in range(N): trainer.predict(...)` with a **single predict pass** over a dataset that's been expanded to length `N × len(base_dataset)`, with each base item contributing `N` consecutive rows. Same-subject rows are guaranteed adjacent, so the dataloader's batches contain tight groups of same-subject copies.

Per-row metadata — `(subject_idx, sample_idx)` — rides through the batch so that `format_trajectories` and the parquet writer can unscramble the flat predictions back into the per-sample output files.

### Dataset wrapper

```python
# src/MEDS_EIC_AR/generation/repeated_dataset.py  (new)
class RepeatedPredictionDataset(torch.utils.data.Dataset):
    """Wraps a base dataset so each underlying item contributes `n_samples` consecutive rows.

    The ordering is (subject_idx changes slow, sample_idx changes fast), so a batch of
    ``batch_size`` covers at most ``ceil(batch_size / n_samples)`` distinct subjects and at least
    ``floor(batch_size / n_samples)`` full sample-groups.

    Concretely: if n_samples=4 and the base dataset is [A, B, C, ...], this dataset yields
        A#0 A#1 A#2 A#3 B#0 B#1 B#2 B#3 C#0 ...
    with each `A#k` carrying metadata ``(subject_idx=0, sample_idx=k)``.
    """

    def __init__(self, base: Dataset, n_samples: int):
        if n_samples < 1:
            raise ValueError(f"n_samples must be >= 1, got {n_samples}")
        self.base = base
        self.n_samples = n_samples

    def __len__(self) -> int:
        return len(self.base) * self.n_samples

    def __getitem__(self, idx: int):
        subject_idx, sample_idx = divmod(idx, self.n_samples)
        item = self.base[subject_idx]
        # The base item is an opaque object (MEDSTorchBatch-compatible dict or dataclass).
        # We attach metadata as a sidecar; the collate_fn repacks it onto the batch.
        return item, subject_idx, sample_idx
```

### Collate

The base `MEDSPytorchDataset.collate` already knows how to collate its native items. We wrap it:

```python
def collate_with_meta(raw_items, base_collate):
    items, subject_idxs, sample_idxs = zip(*raw_items)
    batch = base_collate(list(items))
    # MEDSTorchBatch is a dataclass; extend with sidecar tensors for the predict path.
    batch.subject_idxs = torch.as_tensor(subject_idxs, dtype=torch.long)
    batch.sample_idxs = torch.as_tensor(sample_idxs, dtype=torch.long)
    return batch
```

If `MEDSTorchBatch` doesn't accept extra attributes cleanly, carry the metadata via a simple `(batch, subject_idxs, sample_idxs)` tuple and unpack at `predict_step`.

### Module / CLI changes

```python
# src/MEDS_EIC_AR/training/module.py
class MEICARModule(L.LightningModule):
    ...
    def predict_step(self, batch, batch_idx: int, dataloader_idx: int = 0):
        # Per-batch seeding. Reproducibility is now at batch granularity rather than
        # per-sample — see §Seeding tradeoffs below.
        if self._predict_seed_fn is not None:
            seed_everything(self._predict_seed_fn(batch_idx), workers=False)
        tokens = self.model.generate(batch, **self.generation_kwargs)
        return {
            "tokens": tokens,                         # [B, T]
            "subject_idxs": batch.subject_idxs,       # [B]
            "sample_idxs": batch.sample_idxs,         # [B]
        }
```

```python
# src/MEDS_EIC_AR/__main__.py::generate_trajectories
for split in inference.generate_for_splits:
    base_loader = ...
    base_dataset = base_loader.dataset
    N = inference.N_trajectories_per_task_sample

    expanded = RepeatedPredictionDataset(base_dataset, N)
    loader = DataLoader(
        expanded,
        batch_size=base_loader.batch_size,
        shuffle=False,
        num_workers=base_loader.num_workers,
        collate_fn=partial(collate_with_meta, base_collate=base_dataset.collate),
        pin_memory=base_loader.pin_memory,
    )

    M._predict_seed_fn = lambda batch_idx, s=split: hash_based_seed(
        cfg.get("seed", None), s, batch_idx
    )

    predictions = trainer.predict(model=M, dataloaders=loader)

    # Reshape flat [batch] predictions back into per-sample_idx groups and write
    # one parquet per sample.
    by_sample: dict[int, list] = defaultdict(list)
    for pred in predictions:
        for i in range(pred["tokens"].shape[0]):
            by_sample[int(pred["sample_idxs"][i])].append({
                "subject_idx": int(pred["subject_idxs"][i]),
                "tokens": pred["tokens"][i],
            })

    for sample_idx, rows in by_sample.items():
        out_fp = Path(cfg.output_dir) / split / f"{sample_idx}.parquet"
        out_fp.parent.mkdir(parents=True, exist_ok=True)
        if out_fp.is_file() and not cfg.do_overwrite:
            continue
        predictions_df = format_trajectories_from_rows(base_dataset, rows)
        pa_table = GeneratedTrajectorySchema.align(predictions_df.to_arrow())
        pq.write_table(pa_table, out_fp)
```

`format_trajectories_from_rows` is a small shim that adapts the existing `format_trajectories` to the new flat-row shape; in practice it may just be a loop into the existing function with the `(subject_idx, tokens)` pairs for one sample.

## The padding argument, quantified

Given a batch of size `B` where each row has length `L_i`:

- **Current behavior (heterogeneous batches):** pad-to-max per batch. Wasted padding tokens per batch ≈ `B · max_i(L_i) - sum_i(L_i)`. If subject lengths vary significantly within the split (which is normal — some patients have short histories, some long), this waste can approach 50%+ of batch compute on short rows.
- **With sample-grouped batches and `batch_size = k · N` for integer `k`:** each batch contains `k` distinct subjects × `N` copies. The `N` copies are identical in length, so waste is `B · max_i(L_i) - sum_{k distinct}(N · L_i)`. When `k = 1` (`batch_size = N`), waste is **exactly zero**: every row in the batch has the same length.
- **Even without tuning `batch_size` to a multiple of `N`:** adjacent-subject clustering still reduces length variance within each batch (you're averaging over fewer distinct subjects), so max-length-per-batch tightens.

For workloads where `N_trajectories_per_task_sample` is `>= 8` (research configs) and subject lengths are long-tailed (realistic), this is a meaningful GPU-utilization win even on HF with no other changes.

## Prefix cache argument (for #87)

On a prefix-caching backend (vLLM / SGLang), same-subject rows within a batch share KV-cache entries for the prompt prefix. Generation of `N` samples from one subject therefore:

- Pays the prompt-prefill cost once, not `N` times.
- Pays the decode cost `N` times (unavoidable, samples diverge after the first sampled token).

On a full-size model where prefill dominates decode, this is approximately a **`(1 + (N-1)/N) × speedup ≈ N×`** reduction in prefill cost, or equivalently a large fraction of total generation cost.

**None of this win is accessible under the current looped-predict design**, because each pass creates a fresh batch and the cross-pass cache reuse that SGLang's radix tree could in principle offer is blocked by the outer loop's batch reshuffling. The batching-order change in this issue unblocks it.

## Seeding tradeoffs

The current design gives **per-sample reproducibility**: fixing `cfg.seed` and `(split, sample)` produces the same trajectory for a given subject every run.

The proposed design gives **per-batch reproducibility**: fixing `cfg.seed`, the split, and the batch index produces the same batch of trajectories. Row ordering within a batch is deterministic, so the per-(subject, sample) mapping is also stable *given* the batch layout — but it depends on `batch_size` and dataset ordering in a way the current design doesn't.

Two mitigations, both optional:

- **Mitigation A (sufficient for research use):** document the new contract. Seeds are reproducible as long as `batch_size`, `N_trajectories_per_task_sample`, and the dataset itself are unchanged. If any of those change, the same `(subject, sample_idx)` pair may map to a different trajectory. This is usually fine.
- **Mitigation B (stronger, more work):** use backend-level per-request seeds. vLLM and SGLang both accept `SamplingParams(seed=...)` per request, so we can compute `hash_based_seed(cfg.seed, split, subject_idx, sample_idx)` per row and pass it through. HF does not cleanly support per-row seeding in a single `generate()` call, so on the HF backend Mitigation B would require grouping rows by seed and running multiple sub-batches — negating most of the padding win. Recommendation: adopt Mitigation A as the default, and add Mitigation B as an opt-in flag only for the non-HF backends once #87 lands.

## Interaction with existing issues

- **Independent of #87.** No backend changes required. HF-only users still benefit (padding and setup amortization).
- **Prerequisite multiplier for #87.** Without this change, SGLang / vLLM can't reuse KV across samples of the same subject, so the prefix-cache win in #87 is mostly vestigial.
- **Compatible with #86 rolling generation.** The rolling loop runs per-batch-row independently; grouping N same-subject rows into one batch changes nothing about the rolling logic. It just means the rolling loop processes `k · N` rows per batch instead of `k` rows across `N` separate passes.
- **Neutral to #82 time-budget stopping.** The time budget is per-row state; grouping affects throughput, not the stopping condition.
- **Neutral to #53 generation speed logging.** If anything, this makes speed logging more meaningful because tokens-per-second is no longer diluted by the outer-loop overhead.

## Tests

- **Unit test on `RepeatedPredictionDataset`:** length, indexing, metadata attachment. Verify `(subject_idx, sample_idx)` round-trip.
- **Unit test on `collate_with_meta`:** sidecar tensors present and correctly shaped.
- **Integration test (new end-to-end):** replace the existing `generated_trajectories` fixture's CLI invocation with the new code path. Assert the same output structure: one parquet per `(split, sample_idx)` with the right subjects and non-empty data. Assert that the test passes with `N_trajectories_per_task_sample > 1` — which is where the restructuring actually matters.
- **Parity smoke test:** run the new path with `N=1`; it should produce output shape-identical to the old single-pass loop. (Exact token parity is not guaranteed because the seeding granularity changes, but structural parity is.)
- **Padding assertion (optional):** construct a batch with variable-length subjects and verify that with the grouped path, within-batch max-length variance is ≤ the per-subject length variance. This documents the padding win.

## Implementation checklist

- [ ] `src/MEDS_EIC_AR/generation/repeated_dataset.py` with `RepeatedPredictionDataset` and `collate_with_meta`.
- [ ] `MEICARModule.predict_step` returns `{"tokens", "subject_idxs", "sample_idxs"}` and accepts a `_predict_seed_fn`.
- [ ] `__main__.py::generate_trajectories` replaces the `for sample in range(N)` loop with one predict pass over the expanded dataset, then reshapes into per-sample parquets.
- [ ] `format_trajectories` adapted or wrapped to accept the new flat-row shape.
- [ ] Unit tests for the dataset + collate.
- [ ] End-to-end test mirroring the existing `test_generate_trajectories_runs` but using the new path.
- [ ] Documentation note in `configs/README.md` explaining the new seeding contract.

## Motivation recap

The TL;DR of why this is worth doing even before any backend work:

1. **Padding efficiency.** Tightly-grouped batches → near-zero padding waste → higher effective GPU utilization per batch → faster wall-clock even on the current HF backend.
2. **Setup amortization.** One predict pass instead of `N`; one round of worker spawning, one round of dataloader warmup, one round of seed-everything boilerplate.
3. **Unblocks prefix caching.** This is the change that turns #87's SGLang / vLLM adoption from "nice kernel speedup" into "`N×` generation throughput on research configs."
4. **Better speed telemetry.** #53's speed logging becomes more honest because the outer-loop overhead isn't drowning the true decode throughput.
5. **Zero behavioral regression for `N=1` use cases.** The change degenerates into the current path when there's only one sample per subject.

## Labels
`enhancement`, `priority:high`, `Computational Performance`, `Trajectory Generation / Inference`

**Comment by mmcdermott** [comment]
Collaborator feedback from someone who built ETHOS-ARES (one of the creators of SCOPE/REACH):

> "The radix KV-cache is pretty much perfect for computing multiple completions of the same initial timeline and allows for a really ridiculous number of tasks to be batched together."

This strongly endorses the direction of this issue. The interleaving restructure is the prerequisite that unlocks the largest SGLang/vLLM win — without it, prefix caching can't help because each `trainer.predict()` call is a separate request with no shared prefix context.

This issue is independent of the backend choice (SGLang #88 vs vLLM) but the benefit is multiplicative: interleaving × prefix-caching = O(1) prefill per subject instead of O(N).

**Comment by mmcdermott** [comment]
Delivered by PR #103 (merged into `dev`). The PR body carries `closes #89`; closing manually since dev isn't the default branch. `RepeatedPredictionDataset` interleaves N samples per subject into a single predict pass; downstream demux in `__main__.generate_trajectories` splits the batched output back into per-trajectory parquets.
